graylog2-server/src/main/java/org/graylog2/outputs/IndexSetAwareMessageOutputBuffer.java

```diff
@@ -81,12 +101,15 @@ public void appendAndFlush(FilteredMessage filteredMessage, Consumer<List<Filter
         synchronized (this) {
             // See class the class documentation for the reasoning behind the bufferLength calculation.
             buffer.add(filteredMessage);
-            bufferLength.addAndGet(Math.max(filteredMessage.message().getIndexSets().size(), 1));
+            bufferLength += Math.max(filteredMessage.message().getIndexSets().size(), 1);
+            bufferSizeBytes += estimateOsBulkRequestSize(filteredMessage.message(), objectMapper);
 
-            if (bufferLength.get() >= maxBufferSize) {
+            if ((maxBufferSizeBytes != 0L && bufferSizeBytes >= maxBufferSizeBytes) ||
+                    maxBufferSizeCount != 0 && bufferLength >= maxBufferSizeCount) {
                 flushBatch = buffer;
-                buffer = new ArrayList<>(maxBufferSize);
-                bufferLength.set(0);
+                buffer = new ArrayList<>(bufferLength);
```

Comment:
The idea here is to use the last buffer length as the new buffer length?

Reply:
Yes. We don't have a good value for the initial size. We start with `500` just because we have to start with something and then we'll use the last buffer length because that's hopefully close enough to what we'll end up with for the next buffer.